src/Components/Components/src/Routing/TemplateSegment.cs

```diff
@@ -25,6 +39,10 @@ public TemplateSegment(string template, string segment, bool isParameter)
                     throw new ArgumentException($"Malformed parameter '{segment}' in route '{template}' has no name before the constraints list.");
                 }
 
+                // Set the IsOptional flag to true if any type constraints
+                // for this parameter are designated as optional.
+                IsOptional = tokens.Skip(1).Any(token => token.Contains('?'));
```

Comment:
Quite true. To resolve this and the other point you made above, I'll  move the guard that ensures the `?` appears at the end to the top of the constructor so it will capture both scenarios.